Desktop: Feat: Add useToast hook and refactor TrashNotification to use it

maggie897 opened the pull request:
**Context**
Currently, the notification system in Joplin often requires creating a custom notification component (e.g., TrashNotification, UpdateNotification) even when the requirement is simply to display a short message. This leads to repetitive code and makes adding simple notifications less efficient.

**Changes**
This PR introduces a reusable useToast hook that provides a generic way to display notifications:
	•	Implemented useToast in packages/app-desktop/gui/hooks/useToast.ts
	•	Refactored TrashNotification to use the new hook, replacing direct popupManager.createPopup calls
	•	The hook supports:
	•	Custom message (React node)
	•	Configurable duration (number | 'persist')
	•	Notification type (Info, Success, etc.)
	•	If duration is provided, the toast automatically dismisses after the timeout.

**Result**
The TrashNotification now works exactly as before but is implemented with the new useToast hook. This paves the way for other notifications to be migrated in the future.

Related Issue: #12732

laurent22 commented:
Please address the message in CI:

Error: .gitignore or .eslintignore would be modified - run `cd "/home/runner/work/joplin/joplin" && yarn updateIgnored`
    at main (/home/runner/work/joplin/joplin/packages/tools/checkIgnoredFiles.js:22:15)

maggie897 commented:
Thanks for pointing this out! I’ve run yarn updateIgnored and committed the updated .gitignore / .eslintignore files. Please let me know if anything else is needed.

laurent22 commented on packages/app-desktop/gui/hooks/useToast.ts:
```diff
@@ -0,0 +1,31 @@
+	return useCallback(
+		(
+			message: React.ReactNode,
+			duration: Duration = 4000,
```
Please change the API to `message: React.ReactNode | string, options: ToastOptions = null`

ToastOptions should have your options: `delay` (instead of duration, defaults to 4000), `type`, `persistent` (defaults to `false`).

laurent22 reviewed (commented):
Thank you for working on this! Would it make sense to update all other trash notification messages to verify that your implementation works in all cases?

maggie897 commented:
Hi Laurent, I’ve completed the following updates:
• Refactored useToast hook to standardize toast usage
• Migrated both TrashNotification and ConversionNotification to use the new toast hook
• Verified that both are functional and display correct toasts as expected

Next, I’ll look into updating UpdateNotification and PluginUpdateNotification similarly.
Let me know if you have any suggestions!

Thanks

laurent22 commented on packages/app-desktop/gui/hooks/useToast.ts:
```diff
@@ -0,0 +1,32 @@
+				persistent = false,
+			} = options;
+
+			const handle: PopupHandle = popupManager.createPopup(() => message, { type });
```
Please don't define types that can inferred - https://joplinapp.org/help/dev/coding_style/#dont-set-the-type-when-it-can-be-inferred

laurent22 commented on packages/app-desktop/gui/hooks/useToast.ts:
```diff
@@ -0,0 +1,32 @@
+					handle.scheduleDismiss?.(delay);
+				} catch {
+					handle.scheduleDismiss?.();
+				}
```
- Why a try/catch block here - do you mean that when providing a delay it can fail, but it doesn't without it?
- Why the `?`. Does that mean that `handle` doesn't always have a `scheduleDismiss` method?

maggie897 commented:
Thanks for the feedback!
1. try/catch: I initially added it as a safeguard in case scheduleDismiss(duration) threw unexpectedly (e.g. from a plugin), but you’re right—it’s not needed. I’ve removed it.
2. Optional chaining (?.): Also removed. It was overly defensive—scheduleDismiss should always exist on the popup handle.
3. Inferred types: removed as well.

Also renamed the hook usage to notify = useToastNotifier() to avoid confusion with the existing toast state object, and to better reflect its function as a trigger.

laurent22 commented on packages/app-desktop/gui/ConversionNotification/ConversionNotification.tsx:
```diff
@@ -1,28 +1,31 @@
 import { Dispatch } from 'redux';
-import { PopupNotificationContext } from '../PopupNotification/PopupNotificationProvider';
 import { NotificationType } from '../PopupNotification/types';
+import useToastNotifier from '../../gui/hooks/useToast';
```
Either you call it useToastNotifier or you call it useToast, but you need to pick one as we can't have both. Second time I'm giving you a link to the coding style which you should have read:

https://joplinapp.org/help/dev/coding_style/#use-the-same-case-for-imported-and-exported-members

laurent22 commented on packages/app-desktop/gui/TrashNotification/TrashNotification.tsx:
```diff
@@ -51,11 +51,15 @@ export default (props: Props) => {
+			{
+				delay: 4000,
+				type: NotificationType.Success,
+				persistent: false,
+			},
+		);
+	}, [props.lastDeletion, props.dispatch, notify]);
```
I can't help but notice that this actually the same amount of code, or maybe more than the original one. And this PR is meant to simplify notifications.

See the original issue:

> So we should make it possible to call a simple function that would display a notification. Maybe based on the existing plugin dialogs.showToast() function?

I think this is what we need

maggie897 commented:
Hi Laurent, thanks for the reminder.

I’ve now updated all relevant files to use the unified useToastNotifier hook.
I also changed the default type to Success, so both trash and conversion messages no longer need to specify the type explicitly — they can just call notify(message) directly. This helped reduce some code redundancy and keeps things a bit cleaner.

Let me know if you’d like any further adjustments!

laurent22 commented:
Thanks for updating this @maggie897, that looks much better now. Setting the default to Success also makes sense and simplify the code further.

Have you considered applying the change to PluginNotification.tsx and UpdateNotification.tsx too? Those are more complex but at least the `popupManager.createPopup` can be changed to your new function I think

laurent22 commented:
Please let me know if you're still interested in completing this PR

laurent22 commented:
Closed stale PR
